### PR TITLE
Improve context switch route. Enable cache hash event

### LINKED ISSUE #5084: [Github] Improve context switch route. Enable cache hash event
### 1. Why is this change necessary?
In many projects it is necessary to store additional parameters in the "session" of a user. However, the route only supports the Shopware core parameters. This change adds an event to allow additional parameters via plugin.
Furthermore, the event for building the cache hash is already activated, since the parameters are 1:1 with the current parameters.

---

Imported from Github. Please see: https://github.com/shopware/shopware/pull/4946

### PR DESCRIPTION
**Pull request by OliverSkroblin**
### 1. Why is this change necessary?
In many projects it is necessary to store additional parameters in the "session" of a user. However, the route only supports the Shopware core parameters. This change adds an event to allow additional parameters via plugin.
Furthermore, the event for building the cache hash is already activated, since the parameters are 1:1 with the current parameters.

### CODE REVIEW

File: src/Core/System/SalesChannel/Context/SalesChannelContextService.php [hunk 1]
```diff
@@ -92,7 +92,7 @@ public function get(SalesChannelContextServiceParameters $parameters): SalesChan
             }
 
             $context = $this->factory->create($token, $parameters->getSalesChannelId(), $session);
-            $this->eventDispatcher->dispatch(new SalesChannelContextCreatedEvent($context, $token));
+            $this->eventDispatcher->dispatch(new SalesChannelContextCreatedEvent($context, $token, $session. $parameters));
```
Comment (AydinHassan) [inline src/Core/System/SalesChannel/Context/SalesChannelContextService.php]:
```suggestion
            $this->eventDispatcher->dispatch(new SalesChannelContextCreatedEvent($context, $token, $session, $parameters));
```

**Review by AydinHassan — CHANGES_REQUESTED** [review]

File: src/Core/System/SalesChannel/SalesChannel/ContextSwitchRoute.php [hunk 2]
```diff
@@ -76,6 +77,10 @@ public function switchContext(RequestDataBag $data, SalesChannelContext $context
             ->add(self::STATE_ID, new Type('string'))
         ;
 
+        $event = new SwitchContextEvent($data, $definition, $parameters, $context);
```
Comment (mitelg) [inline src/Core/System/SalesChannel/SalesChannel/ContextSwitchRoute.php:80]: would be nice, if you could use our new way of dispatching events https://github.com/shopware/shopware/blob/trunk/adr/2024-06-18-extended-event-system.md

you might probably hear of this :wink:

File: src/Core/System/SalesChannel/Event/SalesChannelContextCreatedEvent.php [hunk 3]
```diff
@@ -12,21 +12,32 @@
 class SalesChannelContextCreatedEvent extends Event implements ShopwareSalesChannelEvent
 {
     public function __construct(
-        private readonly SalesChannelContext $salesChannelContext,
-        private readonly string $usedToken
+        public readonly SalesChannelContext $context,
```
Comment (mitelg) [inline src/Core/System/SalesChannel/Event/SalesChannelContextCreatedEvent.php:15]: please do not change the mechanics of the event here yet. this should be done as an own PR and for all events and probably along with an ADR.
it would be strange, if some events behave like this and some don't

File: src/Core/Framework/Adapter/Cache/Http/CacheResponseSubscriber.php [hunk 4]
```diff
@@ -223,36 +223,26 @@ private function hasInvalidationState(array $cacheStates, array $states): bool
 
     private function buildCacheHash(Request $request, SalesChannelContext $context): string
     {
-        if (Feature::isActive('cache_rework')) {
```
Comment (mitelg) [inline src/Core/Framework/Adapter/Cache/Http/CacheResponseSubscriber.php:226]: is it okay to directly use it without the feature activated? 🤔

File: src/Core/System/SalesChannel/Event/SwitchContextEvent.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace Shopware\Core\System\SalesChannel\Event;
+
+use Shopware\Core\Framework\Validation\DataBag\RequestDataBag;
+use Shopware\Core\Framework\Validation\DataValidationDefinition;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+
+class SwitchContextEvent
```
Comment (mitelg) [inline src/Core/System/SalesChannel/Event/SwitchContextEvent.php:9]: please implement the ShopwareEventInterface and getters like in other events for now. see comment above

**Review by mitelg — CHANGES_REQUESTED** [review]
would be nice, if you could add some tests and fix the PHPStan and code style issues 🙂👍

File: src/Core/System/SalesChannel/SalesChannel/ContextSwitchRoute.php (hunk 2 above)
Comment (OliverSkroblin) [reply src/Core/System/SalesChannel/SalesChannel/ContextSwitchRoute.php:80]: IMO this is not correct. The new system is designed for pre / post / replace functions. But here you just want a "notify"

File: src/Core/System/SalesChannel/Event/SalesChannelContextCreatedEvent.php (hunk 3 above)
Comment (OliverSkroblin) [reply src/Core/System/SalesChannel/Event/SalesChannelContextCreatedEvent.php:15]: Don't get what you mean. I justed changed the visibility and provided the session data. If you don't want to move it in this direction it would be very polite to just import the PR and revert this change.

File: src/Core/Framework/Adapter/Cache/Http/CacheResponseSubscriber.php (hunk 4 above)
Comment (OliverSkroblin) [reply src/Core/Framework/Adapter/Cache/Http/CacheResponseSubscriber.php:226]: Yes, right now the outcome of the hash is the "public api" (which is not really extendable). So right now just the values are considered to be a break when they would change - but this is not the case here. So here i just activated the new config and the new event

**Comment by mitelg** [comment]
thanks for your contribution @OliverSkroblin 👍🎉💙
